[LBSE] Implement support for gradients

nikolaszimmermann opened the pull request:
#### 1ab7013b94f94e3e31c7c1776cc9adfa34e3ff06
<pre>
[LBSE] Implement support for gradients
<a href="https://bugs.webkit.org/show_bug.cgi?id=265138">https://bugs.webkit.org/show_bug.cgi?id=265138</a>

Reviewed by Rob Buis.

Implement support for complex &quot;paint servers&quot; in LBSE, aka. non-solid-color filling/stroking.
Introduce RenderSVGResourcePaintServer, as base class for RenderSVGResourceGradient, which
is needed by RenderSVGResourceLinearGradient &amp; RenderSVGResourceRadialGradient.

SVGPaintServerHandling contains all necessary code to request a fill/stroke paint server,
honoring the complex fill/stroke fallback rules (when to inherit from parent style, etc.).

This aligns the invalid fill/stroke URI handling with Gecko &amp; Blink, improving interop.

Previously on Apple platforms, gradients on text were realized by filling a rectangle
using the gradient into a separated ImageBuffer, and then applying this as mask when
painting the filled/stroked text. Rework this code to only use compositing operations
and transparency layers in a cross-platform way.

Following the implementation of the LBSE-style resources, such as RenderSVGResourceClipper
and RenderSVGResourceMasker, LBSE no longer makes use of re-layouts to communicate paint server
changes. Only repaints are necessary - see SVGGradientElement::invalidateGradientResource().

Fixes dozens of tests in LBSE - no changes for the legacy engine.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1-SE/svgdom-over-01-f-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-mask-01-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-07-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-08-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-11-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/as-background-image/svg-as-background-6-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/css/path-gradient-stroke-shadow-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/css/rect-gradient-stroke-shadow-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/css/text-gradient-shadow-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-sized-content-with-resources-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/broken-internal-references-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/dominant-baseline-hanging-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/dominant-baseline-hanging-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/external-paintserver-reference-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/fill-fallback-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-add-stops-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-attr-update-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-cycle-detection-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-deep-referencing-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-rotated-bbox-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-stop-corner-cases-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-stop-style-change-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-stroke-width-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-userSpaceOnUse-with-percentage-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-with-1d-boundingbox-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-with-1d-boundingbox-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-2-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-2-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/group-opacity-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/inline-svg-in-xhtml-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-css-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-fill-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-uri-stroke-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-and-object-creation-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-and-object-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-creation-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-stop-linked-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/large-bounding-box-percents-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/non-scaling-stroke-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/radial-gradient-with-outstanding-focalPoint-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/radialGradient-focal-radius-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/resource-client-removal-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/stroke-fallback-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/text-rotated-gradient-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/text-rotated-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/uri-reference-handling-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/use-multiple-on-nested-disallowed-font-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/xlink-custom-namespace-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/xlink-to-invalid-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/error/003-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/perf/004-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/text/selection-background-color-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/text/selection-styles-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/text/text-gradient-positioning-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/transforms/text-with-mask-with-svg-transform-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/transforms/text-with-mask-with-svg-transform-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/zoom/page/zoom-background-images-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/zoom/page/zoom-mask-with-percentages-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/zoom/page/zoom-mask-with-percentages-expected.txt: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedSVGResourceIDs):
(WebCore::ReferencedSVGResources::updateReferencedResources):
(WebCore::ReferencedSVGResources::elementForResourceIDs):
(WebCore::ReferencedSVGResources::referencedPaintServerElement):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgFillPaintServerResourceFromStyle const):
(WebCore::RenderLayerModelObject::svgStrokePaintServerResourceFromStyle const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourceGradient const):
(WebCore::RenderObject::isRenderSVGResourcePaintServer const):
(WebCore::RenderObject::isRenderSVGResourceLinearGradient const):
(WebCore::RenderObject::isRenderSVGResourceRadialGradient const):
* Source/WebCore/rendering/svg/RenderSVGGradientStop.cpp:
(WebCore::RenderSVGGradientStop::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp: Added.
(WebCore::RenderSVGResourceGradient::RenderSVGResourceGradient):
(WebCore::RenderSVGResourceGradient::stopsByApplyingColorFilter const):
(WebCore::RenderSVGResourceGradient::platformSpreadMethodFromSVGType const):
(WebCore::RenderSVGResourceGradient::buildGradientIfNeeded):
(WebCore::RenderSVGResourceGradient::prepareFillOperation):
(WebCore::RenderSVGResourceGradient::prepareStrokeOperation):
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceGradientInlines.h: Added.
(WebCore::RenderSVGResourceGradient::gradientElement const):
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.cpp: Added.
(WebCore::RenderSVGResourceLinearGradient::RenderSVGResourceLinearGradient):
(WebCore::RenderSVGResourceLinearGradient::collectGradientAttributesIfNeeded):
(WebCore::RenderSVGResourceLinearGradient::createGradient):
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradientInlines.h: Added.
(WebCore::RenderSVGResourceLinearGradient::linearGradientElement const):
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.cpp: Added.
(WebCore::RenderSVGResourcePaintServer::RenderSVGResourcePaintServer):
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradient.cpp: Added.
(WebCore::RenderSVGResourceRadialGradient::RenderSVGResourceRadialGradient):
(WebCore::RenderSVGResourceRadialGradient::collectGradientAttributesIfNeeded):
(WebCore::RenderSVGResourceRadialGradient::createGradient):
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradient.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradientInlines.h: Added.
(WebCore::RenderSVGResourceRadialGradient::radialGradientElement const):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::fillContains):
(WebCore::RenderSVGShape::strokeContains):
(WebCore::RenderSVGShape::fillShape):
(WebCore::RenderSVGShape::strokeShape):
(WebCore::RenderSVGShape::strokeShapeInternal): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::SVGInlineTextBox):
(WebCore::SVGInlineTextBox::paintSelectionBackground):
(WebCore::SVGInlineTextBox::paint):
(WebCore::SVGInlineTextBox::acquirePaintingResource):
(WebCore::SVGInlineTextBox::releasePaintingResource):
(WebCore::SVGInlineTextBox::acquireLegacyPaintingResource):
(WebCore::SVGInlineTextBox::releaseLegacyPaintingResource):
(WebCore::SVGInlineTextBox::paintDecorationWithStyle):
(WebCore::SVGInlineTextBox::paintTextWithShadows):
(WebCore::SVGInlineTextBox::prepareGraphicsContextForTextPainting): Deleted.
(WebCore::SVGInlineTextBox::restoreGraphicsContextAfterTextPainting): Deleted.
* Source/WebCore/rendering/svg/SVGInlineTextBox.h:
* Source/WebCore/rendering/svg/SVGPaintServerHandling.h: Added.
(WebCore::SVGPaintServerHandling::SVGPaintServerHandling):
(WebCore::SVGPaintServerHandling::context const):
(WebCore::SVGPaintServerHandling::preparePaintOperation const):
(WebCore::SVGPaintServerHandling::requestPaintServer):
(WebCore::SVGPaintServerHandling::prepareFillOperation const):
(WebCore::SVGPaintServerHandling::prepareStrokeOperation const):
(WebCore::SVGPaintServerHandling::resolveColorFromStyle):
(WebCore::SVGPaintServerHandling::inheritColorFromParentStyleIfNeeded):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceLinearGradient.cpp:
(WebCore::LegacyRenderSVGResourceLinearGradient::LegacyRenderSVGResourceLinearGradient):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceRadialGradient.cpp:
(WebCore::LegacyRenderSVGResourceRadialGradient::LegacyRenderSVGResourceRadialGradient):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGGradientElement.cpp:
(WebCore::SVGGradientElement::invalidateGradientResource):
(WebCore::SVGGradientElement::svgAttributeChanged):
(WebCore::SVGGradientElement::childrenChanged):
* Source/WebCore/svg/SVGGradientElement.h:
* Source/WebCore/svg/SVGLinearGradientElement.cpp:
(WebCore::SVGLinearGradientElement::svgAttributeChanged):
(WebCore::SVGLinearGradientElement::createElementRenderer):
* Source/WebCore/svg/SVGRadialGradientElement.cpp:
(WebCore::SVGRadialGradientElement::svgAttributeChanged):
(WebCore::SVGRadialGradientElement::createElementRenderer):

Canonical link: <a href="https://commits.webkit.org/272653@main">https://commits.webkit.org/272653@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bca29bcd9b89ff04a785a3d3ac1d7d6257ab0dbe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11281 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35099 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29418 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33401 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13632 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8476 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9509 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29077 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8427 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/29017 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36435 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29440 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34551 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6508 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32411 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10225 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7571 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9140 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/svg/SVGGradientElement.cpp:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2004, 2005, 2006, 2007 Rob Buis <buis@kde.org>
  * Copyright (C) Research In Motion Limited 2010. All rights reserved.
  * Copyright (C) 2018-2019 Apple Inc. All rights reserved.
+ * Copyright (C) 2023 Igalia S.L.
```
I guess we have to include 2024 these days :)

nikolaszimmermann replied:
That code is actually from 2023, but I will remember to add 2024, where appropriate.

nikolaszimmermann replied:
Done, selectively.

rwlbuis commented on Source/WebCore/svg/SVGGradientElement.cpp:
```diff
@@ -80,6 +82,15 @@ void SVGGradientElement::svgAttributeChanged(const QualifiedName& attrName)
+                gradientRenderer->invalidateGradient();
+            return;
+        }
+#endif
```
This section seems repeated, should we have a helper method on SVGGradientElement?

nikolaszimmermann replied:
Good catch, will use it from SVGLinear/SVGRadialGradientElement as well.

rwlbuis reviewed (commented):
Still studying the patch, some early feedback...

rwlbuis commented on Source/WebCore/rendering/ReferencedSVGResources.h:
```diff
@@ -71,10 +74,12 @@ class ReferencedSVGResources {
 
 private:
     static RefPtr<SVGElement> elementForResourceID(TreeScope&, const AtomString& resourceID, const QualifiedName& tagName);
+    static RefPtr<SVGElement> elementForResourceIDs(TreeScope&, const AtomString& resourceID, const QualifiedNames& tagNames);
```
I wonder if we can just always use the QualifiedNames variant?

nikolaszimmermann replied:
Yeah, but the former variant saves the QualifiedNames traversal

rwlbuis commented on Source/WebCore/rendering/ReferencedSVGResources.cpp:
```diff
@@ -110,45 +111,62 @@ Vector<std::pair<AtomString, QualifiedName>> ReferencedSVGResources::referencedS
         oldKeys.remove(targetID);
+
+        for (auto& targetID : oldKeys)
+            removeClientForTarget(treeScope, targetID);
```
Why did this move into the for loop?

nikolaszimmermann replied:
Ouch, probably a merge problem. Fixed.

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.cpp:
```diff
@@ -0,0 +1,72 @@
+    if (m_attributes.has_value())
+        return;
+
+    m_attributes = std::nullopt;
```
Is this needed, since m_attributes has no value at this point?

nikolaszimmermann replied:
Good catch, will remove the nop.

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h:
```diff
@@ -0,0 +1,50 @@
+class RenderSVGResourcePaintServer : public RenderSVGResourceContainer {
+    WTF_MAKE_ISO_ALLOCATED(RenderSVGResourcePaintServer);
+public:
+    virtual ~RenderSVGResourcePaintServer();
```
It seems we sometimes use "= default" for dtors in headers but here we use it in the cpp. I guess we should have a consistent style, I prefer to have it in the header.

nikolaszimmermann replied:
I always keep virtual destructors in the .cpp file - no strong opinion here. Need to refresh my mind about that topic :-)

rwlbuis commented on Source/WebCore/rendering/svg/SVGInlineTextBox.h:
```diff
@@ -23,14 +24,21 @@
+class SVGPaintServerHandling;
 class SVGRootInlineBox;
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
```
I think you can forward reference just about any class, existing or not, so I don't think this part has to be protected.

nikolaszimmermann replied:
Wasn't sure if we had a preferred style regarding that topic. Remove the guard.

rwlbuis commented on Source/WebCore/rendering/svg/SVGInlineTextBox.h:
```diff
@@ -23,14 +24,21 @@
 class SVGRootInlineBox;
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+class RenderSVGResourcePaintServer;
```
I suggest either forward referencing or including, not both.

nikolaszimmermann replied:
Fixed.

rwlbuis commented on Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
```diff
@@ -52,14 +53,15 @@ struct ExpectedSVGInlineTextBoxSize : public LegacyInlineTextBox {
     float float1;
     uint32_t bitfields : 5;
     void* pointer;
+    SVGPaintServerOrColor paintServerOrColor;
```
Does this field need if ENABLE(LAYER_BASED_SVG_ENGINE) ?

nikolaszimmermann replied:
Good catch!

rwlbuis commented on Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
```diff
@@ -554,24 +616,68 @@ void SVGInlineTextBox::paintTextWithShadows(GraphicsContext& context, const Rend
+            // Optimized code path to support gradient fill/stroke on text without using temporary ImageBuffers / masking.
+            auto* fillGradient = paintingResourceMode().contains(RenderSVGResourceMode::ApplyToFill) ? usedContext->fillGradient() : nullptr;
+            auto* strokeGradient = paintingResourceMode().contains(RenderSVGResourceMode::ApplyToStroke) ? usedContext->strokeGradient() : nullptr;
+            if (renderer().document().settings().layerBasedSVGEngineEnabled() && (fillGradient || strokeGradient)) {
```
I think it may work better to set fillGradient/strokeGradient to nullptr and only try assign it a value if layerBasedSVGEngineEnabled() is true. Then below we do not need to check layerBasedSVGEngineEnabled() a second time.

nikolaszimmermann replied:
Good idea, fixed.

weinig commented on Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.cpp:
```diff
@@ -0,0 +1,72 @@
+
+    return Gradient::create(
+        Gradient::LinearData { startPoint, endPoint },
+        { ColorInterpolationMethod::SRGB { }, AlphaPremultiplication::Unpremultiplied },
```
I don't believe I ever ended up doing this in the SVG code, but if you are looking at this, one nice improvement you could look at making would be to support the "color-interpolation" property (https://www.w3.org/TR/SVG2/painting.html#ColorInterpolationProperty) and allow users to use linearSRGB here if they wanted. (Same with radial gradients).

The platform Gradient code knows how draw them, we just never piped it in).

(obviously, not for this PR, just wanted to point it out)

nikolaszimmermann replied:
Great idea, thanks @weinig for the reminder. We should open a ticket to track it.

rwlbuis commented on LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-userSpaceOnUse-with-percentage-expected.txt:
```diff
@@ -0,0 +1,12 @@
+layer at (0,50) size 300x50
+  RenderSVGRect {rect} at (0,50) size 300x50 [fill={[type=SOLID] [color=#000000]}] [x=0.00] [y=50.00] [width=300.00] [height=50.00]
+layer at (0,100) size 300x300
+  RenderSVGRect {rect} at (0,100) size 300x300 [fill={[type=SOLID] [color=#000000]}] [x=0.00] [y=100.00] [width=300.00] [height=300.00]
```
These and others do not refer to gradient pservers, for example the default expectation mentions:
RenderSVGRect {rect} at (0,0) size 300x50 [fill={[type=LINEAR-GRADIENT] [id="gradient1"]}] [x=0.00] [y=0.00] [width=300.00] [height=50.00]

I am not sure we need to dump the pserver itself, but the fill type I guess is needed. WDYT?

nikolaszimmermann replied:
Indeed, the render tree text dumps  do not know anything about the new resources - follow up pr?

rwlbuis replied:
Sure, probably more important to land https://github.com/WebKit/WebKit/pull/22352 first anyway.

rwlbuis commented on Source/WebCore/svg/SVGRadialGradientElement.cpp:
```diff
@@ -98,6 +99,7 @@ void SVGRadialGradientElement::svgAttributeChanged(const QualifiedName& attrName
         InstanceInvalidationGuard guard(*this);
         updateRelativeLengthsInformation();
+        invalidateGradientResource();
         updateSVGRendererForElementChange();
```
I think this call to updateSVGRendererForElementChange is not needed (I commented before, not sure if it was lost).

nikolaszimmermann replied:
Good catch!

rwlbuis reviewed (approved):
Looks good with one potential line removal.

nikolaszimmermann committed:
[LBSE] Implement support for gradients
https://bugs.webkit.org/show_bug.cgi?id=265138

Reviewed by Rob Buis.

Implement support for complex "paint servers" in LBSE, aka. non-solid-color filling/stroking.
Introduce RenderSVGResourcePaintServer, as base class for RenderSVGResourceGradient, which
is needed by RenderSVGResourceLinearGradient & RenderSVGResourceRadialGradient.

SVGPaintServerHandling contains all necessary code to request a fill/stroke paint server,
honoring the complex fill/stroke fallback rules (when to inherit from parent style, etc.).

This aligns the invalid fill/stroke URI handling with Gecko & Blink, improving interop.

Previously on Apple platforms, gradients on text were realized by filling a rectangle
using the gradient into a separated ImageBuffer, and then applying this as mask when
painting the filled/stroked text. Rework this code to only use compositing operations
and transparency layers in a cross-platform way.

Following the implementation of the LBSE-style resources, such as RenderSVGResourceClipper
and RenderSVGResourceMasker, LBSE no longer makes use of re-layouts to communicate paint server
changes. Only repaints are necessary - see SVGGradientElement::invalidateGradientResource().

Fixes dozens of tests in LBSE - no changes for the legacy engine.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1-SE/svgdom-over-01-f-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-mask-01-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-07-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-08-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-11-b-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/as-background-image/svg-as-background-6-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/css/path-gradient-stroke-shadow-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/css/rect-gradient-stroke-shadow-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/css/text-gradient-shadow-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-sized-content-with-resources-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/broken-internal-references-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/dominant-baseline-hanging-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/dominant-baseline-hanging-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/external-paintserver-reference-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/fill-fallback-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-add-stops-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-attr-update-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-cycle-detection-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-deep-referencing-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-rotated-bbox-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-stop-corner-cases-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-stop-style-change-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-stroke-width-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-userSpaceOnUse-with-percentage-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-with-1d-boundingbox-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/gradient-with-1d-boundingbox-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-2-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-2-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/grayscale-gradient-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/group-opacity-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/inline-svg-in-xhtml-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-css-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-fill-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-uri-stroke-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-and-object-creation-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-and-object-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-creation-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-gradient-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-update-stop-linked-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/large-bounding-box-percents-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/non-scaling-stroke-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/radial-gradient-with-outstanding-focalPoint-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/radialGradient-focal-radius-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/resource-client-removal-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/stroke-fallback-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/text-rotated-gradient-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/text-rotated-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/uri-reference-handling-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/use-multiple-on-nested-disallowed-font-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/xlink-custom-namespace-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/xlink-to-invalid-gradient-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/error/003-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/perf/004-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/text/selection-background-color-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/text/selection-styles-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/text/text-gradient-positioning-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/transforms/text-with-mask-with-svg-transform-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/transforms/text-with-mask-with-svg-transform-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/zoom/page/zoom-background-images-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/zoom/page/zoom-mask-with-percentages-expected.png: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/zoom/page/zoom-mask-with-percentages-expected.txt: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedSVGResourceIDs):
(WebCore::ReferencedSVGResources::updateReferencedResources):
(WebCore::ReferencedSVGResources::elementForResourceIDs):
(WebCore::ReferencedSVGResources::referencedPaintServerElement):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgFillPaintServerResourceFromStyle const):
(WebCore::RenderLayerModelObject::svgStrokePaintServerResourceFromStyle const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourceGradient const):
(WebCore::RenderObject::isRenderSVGResourcePaintServer const):
(WebCore::RenderObject::isRenderSVGResourceLinearGradient const):
(WebCore::RenderObject::isRenderSVGResourceRadialGradient const):
* Source/WebCore/rendering/svg/RenderSVGGradientStop.cpp:
(WebCore::RenderSVGGradientStop::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp: Added.
(WebCore::RenderSVGResourceGradient::RenderSVGResourceGradient):
(WebCore::RenderSVGResourceGradient::stopsByApplyingColorFilter const):
(WebCore::RenderSVGResourceGradient::platformSpreadMethodFromSVGType const):
(WebCore::RenderSVGResourceGradient::buildGradientIfNeeded):
(WebCore::RenderSVGResourceGradient::prepareFillOperation):
(WebCore::RenderSVGResourceGradient::prepareStrokeOperation):
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceGradientInlines.h: Added.
(WebCore::RenderSVGResourceGradient::gradientElement const):
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.cpp: Added.
(WebCore::RenderSVGResourceLinearGradient::RenderSVGResourceLinearGradient):
(WebCore::RenderSVGResourceLinearGradient::collectGradientAttributesIfNeeded):
(WebCore::RenderSVGResourceLinearGradient::createGradient):
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradientInlines.h: Added.
(WebCore::RenderSVGResourceLinearGradient::linearGradientElement const):
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.cpp: Added.
(WebCore::RenderSVGResourcePaintServer::RenderSVGResourcePaintServer):
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradient.cpp: Added.
(WebCore::RenderSVGResourceRadialGradient::RenderSVGResourceRadialGradient):
(WebCore::RenderSVGResourceRadialGradient::collectGradientAttributesIfNeeded):
(WebCore::RenderSVGResourceRadialGradient::createGradient):
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradient.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradientInlines.h: Added.
(WebCore::RenderSVGResourceRadialGradient::radialGradientElement const):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::fillContains):
(WebCore::RenderSVGShape::strokeContains):
(WebCore::RenderSVGShape::fillShape):
(WebCore::RenderSVGShape::strokeShape):
(WebCore::RenderSVGShape::strokeShapeInternal): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::SVGInlineTextBox):
(WebCore::SVGInlineTextBox::paintSelectionBackground):
(WebCore::SVGInlineTextBox::paint):
(WebCore::SVGInlineTextBox::acquirePaintingResource):
(WebCore::SVGInlineTextBox::releasePaintingResource):
(WebCore::SVGInlineTextBox::acquireLegacyPaintingResource):
(WebCore::SVGInlineTextBox::releaseLegacyPaintingResource):
(WebCore::SVGInlineTextBox::paintDecorationWithStyle):
(WebCore::SVGInlineTextBox::paintTextWithShadows):
(WebCore::SVGInlineTextBox::prepareGraphicsContextForTextPainting): Deleted.
(WebCore::SVGInlineTextBox::restoreGraphicsContextAfterTextPainting): Deleted.
* Source/WebCore/rendering/svg/SVGInlineTextBox.h:
* Source/WebCore/rendering/svg/SVGPaintServerHandling.h: Added.
(WebCore::SVGPaintServerHandling::SVGPaintServerHandling):
(WebCore::SVGPaintServerHandling::context const):
(WebCore::SVGPaintServerHandling::preparePaintOperation const):
(WebCore::SVGPaintServerHandling::requestPaintServer):
(WebCore::SVGPaintServerHandling::prepareFillOperation const):
(WebCore::SVGPaintServerHandling::prepareStrokeOperation const):
(WebCore::SVGPaintServerHandling::resolveColorFromStyle):
(WebCore::SVGPaintServerHandling::inheritColorFromParentStyleIfNeeded):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceLinearGradient.cpp:
(WebCore::LegacyRenderSVGResourceLinearGradient::LegacyRenderSVGResourceLinearGradient):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceRadialGradient.cpp:
(WebCore::LegacyRenderSVGResourceRadialGradient::LegacyRenderSVGResourceRadialGradient):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGGradientElement.cpp:
(WebCore::SVGGradientElement::invalidateGradientResource):
(WebCore::SVGGradientElement::svgAttributeChanged):
(WebCore::SVGGradientElement::childrenChanged):
* Source/WebCore/svg/SVGGradientElement.h:
* Source/WebCore/svg/SVGLinearGradientElement.cpp:
(WebCore::SVGLinearGradientElement::svgAttributeChanged):
(WebCore::SVGLinearGradientElement::createElementRenderer):
* Source/WebCore/svg/SVGRadialGradientElement.cpp:
(WebCore::SVGRadialGradientElement::svgAttributeChanged):
(WebCore::SVGRadialGradientElement::createElementRenderer):

Canonical link: https://commits.webkit.org/272653@main
